src/routes/github/github-router.test.ts

```diff
@@ -0,0 +1,173 @@
+import { when } from "jest-when";
+import { GithubConfigurationGet } from "./configuration/github-configuration-get";
+import { getGitHubApiUrl } from "utils/get-github-client-config";
+import axios from "axios";
+import { GitHubServerApp } from "models/github-server-app";
+import { Installation } from "models/installation";
+import { v4 as v4uuid } from "uuid";
+import { envVars } from "config/env";
+
+jest.mock("./configuration/github-configuration-get");
+jest.mock("utils/get-github-client-config");
+jest.mock("axios", () => ({ get: jest.fn(), create: jest.fn() }));
+jest.mock("models/github-server-app");
+jest.mock("models/installation");
+
+const JIRA_HOST = "https://gh4j.test.atlassian.net";
+const VALID_TOKEN = "valid-token";
+const GITHUB_SERVER_APP_UUID: string = v4uuid();
+const GITHUB_SERVER_APP_ID = 789;
+const GITHUB_SERVER_APP_URL = "https://default.internal.github.atlassian.net";
+const JIRA_INSTALLATION_ID = 444444;
+
+const getGitHubServerAppModel = (): GitHubServerApp => {
+	return {
+		id: 1,
+		uuid: GITHUB_SERVER_APP_UUID,
+		appId: GITHUB_SERVER_APP_ID,
+		gitHubBaseUrl: GITHUB_SERVER_APP_URL,
+		gitHubClientId: "client-id",
+		gitHubClientSecret: "gitHubClientSecret",
+		webhookSecret: "webhookSecret",
+		privateKey: "privateKey",
+		gitHubAppName: "test-app-name",
+		installationId: JIRA_INSTALLATION_ID,
+		updatedAt: new Date(),
+		createdAt: new Date(),
+		decrypt: async (s) => s
+	} as GitHubServerApp;
+};
+
+const getServerApptInstallationModel = (): Installation => {
+	return {
+		id: JIRA_INSTALLATION_ID,
+		jiraHost: JIRA_HOST,
+		secrets: "secrets",
+		sharedSecret: "sharedSecret",
+		encryptedSharedSecret: "encryptedSharedSecret",
+		clientKey: "clientKey",
+		updatedAt: new Date(),
+		createdAt: new Date(),
+		githubAppId: GITHUB_SERVER_APP_ID
+	} as Installation;
+};
+
+const mockPrerequistApp = () => {
+	const app = express();
+	app.use((req, res, next) => {
+		req.session = { githubToken: VALID_TOKEN };
+		res.locals = { jiraHost: JIRA_HOST };
+		req.log = getLogger("test");
+		next();
+	});
+	return app;
+};
+
+const mockGetGitHubApiUrl = () => {
+	when(getGitHubApiUrl)
+		.calledWith(JIRA_HOST, GITHUB_SERVER_APP_ID)
+		.mockResolvedValue(GITHUB_SERVER_APP_URL);
+};
+
+const mockGetForGitHubServerAppId = () => {
+	when(GitHubServerApp.findForUuid)
+		.calledWith(GITHUB_SERVER_APP_UUID)
+		.mockResolvedValue(getGitHubServerAppModel());
+};
+
+const mockAxiosPingURL = () => {
+	when(axios.get)
+		.calledWith(JIRA_HOST)
+		.mockResolvedValue(undefined);
+};
+
+const mockInstallationFindByPK = () => {
+	when(Installation.findByPk)
+		.calledWith(JIRA_INSTALLATION_ID as any)
+		.mockResolvedValue(getServerApptInstallationModel());
+};
+
+const mockConfigurationGetProceed = ()=>{
+	jest.mocked(GithubConfigurationGet).mockClear();
+	jest.mocked(GithubConfigurationGet).mockImplementation(async (_req, res) => {
+		res.end("ok");
+	});
+};
+
+describe("GitHub router", () => {
```

Comment:
this whole suite of test is only testing the middleware from what I can tell and not the route handler, which means that the route handler dependency to `res.locals.githubAppId` or `res.locals.githubAppConfig` is not tested.

Reply:
So the changes to the `github-route.ts` is just added an optional `/:uuid`, therefore this unit test is to test that `/:uuid` is:
1. Optional --> the first "Cloud scenario"
2. When present, uuid is picked up and parsed --> I use one `GithubConfigurationGet` route to test it and assert the locals are expected.

Reply:
I understand that, what I'm saying is that this is a test for the router which should be tested thoroughly, which includes the route handler itself.  What you seem to be doing is testing the router and route handler  separately, which means that there's no test for them working correctly together.

When it comes to testing routes, integration tests are the only way to get full confidence.  You can unit test route handlers for input/output, but you cannot do this with routers without mocking important parts of the flow which may change in the future and break in production.

Reply:
Please review, this is the discussed integration test. But only testing new features in this PR.

Reply:
And I didn't use the real `GitHubConfigurationGet` and assert on the handlebars output, as I think that would be too much for the test.